// @azphalt/azp/dist/trust.js
/**
 * The azphalt trust model (`spec/package-format.md` § Signing). {@link verifyAzp} answers "is this
 * signature internally consistent?" — tamper-evidence. This answers the *identity* question: "does
 * the signer's key belong to someone this host trusts?"
 *
 * A host holds a {@link TrustStore} of trusted Ed25519 public keys. A package is **trusted** when
 * either (a) its signer key is directly in the store, or (b) its signer key was counter-signed by a
 * **registry** key that is in the store — transitive trust, so a host can trust a registry once
 * instead of every author. Key distribution is out-of-band (the host/registry decides); this module
 * only enforces the cryptography.
 */
import { createPrivateKey, createPublicKey, sign as cryptoSign, verify as cryptoVerify } from "node:crypto";
import { strToU8, strFromU8, unzipSync, zipSync } from "fflate";
import { EPOCH, verifyAzp, readAzp } from "./index.js";
/** Ceiling on counter-signature chain length — a DoS guard against attacker-crafted deep chains. */
const MAX_CHAIN_DEPTH = 10;
function parseSignature(azp) {
    const { payload } = readAzp(azp);
    const raw = payload["signature.json"];
    if (!raw)
        return null;
    return JSON.parse(strFromU8(raw));
}
/**
 * Verify a `.azp` against a {@link TrustStore}: first internal consistency (via {@link verifyAzp}),
 * then whether the signer's key is trusted — directly, or through a registry counter-signature by a
 * key in the store.
 */
export function verifyTrust(azp, store) {
    const v = verifyAzp(azp);
    if (!v.ok) {
        return { ok: false, signed: v.signed, trusted: false, reason: `invalid package: ${v.errors.join("; ")}` };
    }
    if (!v.signed) {
        return { ok: true, signed: false, trusted: false, reason: "unsigned: no signer to trust" };
    }
    let sig;
    try {
        sig = parseSignature(azp);
    }
    catch (e) {
        return { ok: true, signed: true, trusted: false, reason: `signature unreadable: ${e.message}` };
    }
    if (!sig)
        return { ok: true, signed: false, trusted: false, reason: "unsigned: no signer to trust" };
    const base = {
        ok: true,
        signed: true,
        signerPublicKey: sig.publicKey,
        signerKeyId: sig.keyId,
    };
    // The store may come from user config; guard against null/malformed input rather than throwing.
    const keys = store?.keys;
    if (!Array.isArray(keys)) {
        return { ...base, trusted: false, reason: "invalid trust store: keys list is missing or malformed" };
    }
    const trustedKeys = new Set(keys.map((k) => k?.publicKey).filter(Boolean));
    // (a) Direct trust: the signer key is in the store.
    if (sig.publicKey && trustedKeys.has(sig.publicKey)) {
        return { ...base, trusted: true, reason: "signer key is directly trusted" };
    }
    // (b) Transitive trust: walk the counter-signature chain from the author up. Each link's key
    // vouches (signs) for the key below it — the author at the base, then each previous counter-signer.
    // Trusted as soon as a link's key is in the store, provided every hop's signature down to it
    // verifies (a broken lower link severs the chain).
    let vouchedKey = sig.publicKey; // base64 SPKI of the key this hop vouches for
    let cs = sig.countersignature;
    let hop = 0;
    while (cs) {
        hop++;
        // Cap the chain: verifying each hop is an Ed25519 verify, so an attacker-crafted deep chain
        // of bad links could burn CPU. A trust chain more than a few hops deep is not a real topology.
        if (hop > MAX_CHAIN_DEPTH) {
            return { ...base, trusted: false, reason: `counter-signature chain exceeds the ${MAX_CHAIN_DEPTH}-hop limit` };
        }
        if (typeof cs.publicKey !== "string" || typeof cs.signature !== "string") {
            return { ...base, trusted: false, reason: `counter-signature is malformed at hop ${hop}` };
        }
        let valid;
        try {
            const pub = createPublicKey({ key: Buffer.from(cs.publicKey, "base64"), format: "der", type: "spki" });
            valid =
                pub.asymmetricKeyType === "ed25519" &&
                    cryptoVerify(null, Buffer.from(vouchedKey, "base64"), pub, Buffer.from(cs.signature, "base64"));
        }
        catch (e) {
            return { ...base, trusted: false, reason: `counter-signature error at hop ${hop}: ${e.message}` };
        }
        if (!valid)
            return { ...base, trusted: false, reason: `counter-signature invalid at hop ${hop}` };
        if (trustedKeys.has(cs.publicKey)) {
            return {
                ...base,
                trusted: true,
                reason: hop === 1 ? "signer counter-signed by a trusted registry" : `signer trusted via a ${hop}-hop counter-signature chain`,
                viaRegistryPublicKey: cs.publicKey,
            };
        }
        vouchedKey = cs.publicKey;
        cs = cs.countersignature;
    }
    return {
        ...base,
        trusted: false,
        reason: sig.countersignature ? "counter-signature chain reaches no trusted key" : "signer key is not in the trust store",
    };
}
/**
 * Counter-sign an already-signed `.azp`: an authority vouches for the current leaf of the trust
 * chain by signing that key's SPKI public-key bytes. The first counter-signature vouches for the
 * author; a further one vouches for the previous counter-signer, extending the chain (web of trust).
 * The author's signature (over `manifest.json`) is untouched, so integrity still verifies. Returns
 * new `.azp` bytes; the input is not mutated.
 */
export function countersign(azp, opts) {
    const files = unzipSync(azp);
    const sigRaw = files["signature.json"];
    if (!sigRaw)
        throw new Error("azp: cannot counter-sign an unsigned package (no signature.json)");
    const sig = JSON.parse(strFromU8(sigRaw));
    if (sig.alg !== "ed25519" || typeof sig.publicKey !== "string") {
        throw new Error("azp: signature.json is malformed");
    }
    const key = createPrivateKey(opts.registryPrivateKey);
    if (key.asymmetricKeyType !== "ed25519")
        throw new Error("azp: registryPrivateKey must be an ed25519 key");
    // Find the deepest existing link; the new counter-signature vouches for its key (the author's if
    // there is no chain yet) and attaches above it.
    let parent = sig;
    let depth = 0;
    while (parent.countersignature) {
        if (typeof parent.countersignature !== "object" || typeof parent.countersignature.publicKey !== "string") {
            throw new Error("azp: existing counter-signature chain is malformed");
        }
        parent = parent.countersignature;
        depth++;
    }
    if (depth >= MAX_CHAIN_DEPTH) {
        throw new Error(`azp: counter-signature chain is already at the ${MAX_CHAIN_DEPTH}-hop limit`);
    }
    const vouchesFor = Buffer.from(parent.publicKey, "base64");
    const signature = cryptoSign(null, vouchesFor, key);
    const registryPublicKey = Buffer.from(createPublicKey(key).export({ type: "spki", format: "der" })).toString("base64");
    parent.countersignature = {
        publicKey: registryPublicKey,
        signature: Buffer.from(signature).toString("base64"),
        ...(opts.keyId ? { keyId: opts.keyId } : {}),
    };
    const entries = {
        ...files,
        "signature.json": strToU8(JSON.stringify(sig, null, 2) + "\n"),
    };
    const sorted = {};
    for (const k of Object.keys(entries).sort())
        sorted[k] = entries[k];
    return zipSync(sorted, { mtime: EPOCH });
}
/**
 * Build a {@link TrustStore} from a list of advertised keys (e.g. a repository's `.well-known`
 * `signingKeys`). This is the trust-bootstrap entry point: a host trusts a registry once, then every
 * author that registry counter-signs is transitively trusted by {@link verifyTrust}. Malformed
 * entries (no string `publicKey`) are dropped rather than throwing.
 */
export function trustStoreFromKeys(keys) {
    const clean = (Array.isArray(keys) ? keys : []).filter((k) => k && typeof k.publicKey === "string");
    return {
        keys: clean.map((k) => ({
            publicKey: k.publicKey,
            keyId: typeof k.keyId === "string" ? k.keyId : undefined,
            label: typeof k.label === "string" ? k.label : undefined,
        })),
    };
}
/** Convenience: a registry key is just an Ed25519 key. Re-exported for symmetry with signing. */
export { generateSigningKey } from "./sign.js";
