// @azphalt/azdk/dist/index.js
/**
 * `@azphalt/azdk` — the typed surface an azphalt extension is written against.
 *
 * Turns the strings in the spec — capabilities, contributions, params, the UI
 * schema — into a typed API. This package is **types plus thin author helpers
 * only**. The real {@link Host} is provided by a conforming runtime (see
 * `packages/runtime-reference`), which injects it and enforces the capability
 * grants. Nothing here can reach a host engine, camera, sensors, the filesystem,
 * or the network — see the never-list in `spec/capability-model.md`.
 */
/** The package-format / manifest version this SDK targets. */
export const FORMAT_VERSION = "0.1";
/** The effective bit depth of a bitmap (absent ⇒ 8). */
export function bitDepth(bitmap) {
    return bitmap.depth ?? 8;
}
/** Bytes per channel for a depth: 1 at 8-bit, 2 at 16-bit. */
export function bytesPerChannel(depth) {
    return depth === 16 ? 2 : 1;
}
/** The maximum channel value for a depth: 255 at 8-bit, 65535 at 16-bit. */
export function maxChannelValue(depth) {
    return depth === 16 ? 65535 : 255;
}
const KIND = Symbol.for("azphalt.contribution");
/**
 * Declare a filter. Export the result under the name your manifest's
 * `contributes.filters[].entry` points at:
 *
 * ```ts
 * import { defineFilter } from "@azphalt/azdk";
 * export const applyHalftone = defineFilter((ctx) => {
 * const cell = ctx.params.number("cellSize");
 * const bmp = ctx.bitmap.read(ctx.target);
 * // …mutate bmp.data (RGBA8, straight alpha)…
 * ctx.bitmap.write(ctx.target, bmp);
 * });
 * ```
 */
export function defineFilter(fn) {
    return Object.assign(fn, { [KIND]: "filter" });
}
export function defineTool(fn) {
    return Object.assign(fn, { [KIND]: "tool" });
}
export function defineCommand(fn) {
    return Object.assign(fn, { [KIND]: "command" });
}
/** Declare a transition (two input frames → one blended output). Branded `"transition"`. */
export function defineTransition(fn) {
    return Object.assign(fn, { [KIND]: "transition" });
}
