// @azphalt/runtime-reference/dist/index.js
/**
 * `@azphalt/runtime-reference` — a reference host that loads a `.azp`, verifies it, builds the
 * host-function table from the granted capabilities, and dispatches a contribution against an
 * in-memory {@link World}. It is the executable form of the contract in `spec/`.
 *
 * **Honest scope.** This first cut exercises the *contract* — capability gating (an ungranted
 * capability is absent, not erroring), the RGBA8 image-buffer ABI, and contribution dispatch —
 * by running a *trusted* extension module in-process. The WASM isolation substrate
 * (QuickJS-in-WASM for `runtime: js`, raw WASM for `runtime: wasm`) that sandboxes untrusted
 * code is the conformance target a production host adds on top; it is intentionally out of scope
 * here. Because there is no sandbox, callers supply the already-imported extension module.
 */
import { readAzp, verifyAzp } from "@azphalt/azp";
import { createHost } from "./host.js";
export * from "./world.js";
export { createHost } from "./host.js";
/** Matches the brand `defineFilter`/`defineTool`/`defineCommand` attach to a contribution. */
const KIND = Symbol.for("azphalt.contribution");
/** Thrown when a `.azp` fails verification or a contribution cannot be dispatched. */
export class AzpError extends Error {
    constructor(message) {
        super(message);
        this.name = "AzpError";
    }
}
/** Verify a `.azp` and parse its manifest + payload. Throws {@link AzpError} if verification fails. */
export function open(azp) {
    const v = verifyAzp(azp);
    if (!v.ok)
        throw new AzpError(`azp verification failed: ${v.errors.join("; ")}`);
    return readAzp(azp);
}
function brandOf(x) {
    if (typeof x !== "function")
        return undefined;
    return x[KIND];
}
function resolve(module, entry, kind) {
    const fn = module[entry];
    if (brandOf(fn) !== kind) {
        throw new AzpError(`export '${entry}' is not a ${kind} contribution (use define${kind[0].toUpperCase()}${kind.slice(1)})`);
    }
    return fn;
}
/**
 * Dispatch a filter from `manifest.contributes.filters` against `world`. Runs the first filter,
 * or the one whose `id` matches `filterId`. The host is built from the granted capabilities only.
 */
export async function runFilter(manifest, module, world, filterId) {
    const filters = manifest.contributes?.filters ?? [];
    const f = filterId ? filters.find((x) => x.id === filterId) : filters[0];
    if (!f)
        throw new AzpError(`no filter${filterId ? ` '${filterId}'` : ""} in manifest.contributes`);
    const fn = resolve(module, f.entry, "filter");
    const host = createHost(manifest, world);
    const active = world.layers.find((l) => l.id === world.activeLayerId);
    if (!active)
        throw new AzpError("world has no active layer");
    // The context is the granted host plus the target layer. Ungranted sub-APIs are absent at
    // runtime (the capability contract); the cast reflects that this is a partial host by design.
    const ctx = { ...host, target: { id: active.id } };
    await fn(ctx);
}
/** Dispatch a tool from `manifest.contributes.tools`. */
export async function runTool(manifest, module, world, toolId) {
    const tools = manifest.contributes?.tools ?? [];
    const t = toolId ? tools.find((x) => x.id === toolId) : tools[0];
    if (!t)
        throw new AzpError(`no tool${toolId ? ` '${toolId}'` : ""} in manifest.contributes`);
    const fn = resolve(module, t.entry, "tool");
    await fn(createHost(manifest, world));
}
/** Dispatch a command from `manifest.contributes.commands`. */
export async function runCommand(manifest, module, world, commandId) {
    const commands = manifest.contributes?.commands ?? [];
    const c = commandId ? commands.find((x) => x.id === commandId) : commands[0];
    if (!c)
        throw new AzpError(`no command${commandId ? ` '${commandId}'` : ""} in manifest.contributes`);
    const fn = resolve(module, c.entry, "command");
    await fn(createHost(manifest, world));
}
/**
 * Dispatch a transition from `manifest.contributes.transitions` against `world`. The two input frames
 * (`from`, `to`) and `progress` (0→1) are supplied by the caller; the transition writes its blended
 * result to the active layer via `bitmap.write`.
 */
export async function runTransition(manifest, module, world, input, transitionId) {
    const transitions = manifest.contributes?.transitions ?? [];
    const t = transitionId ? transitions.find((x) => x.id === transitionId) : transitions[0];
    if (!t)
        throw new AzpError(`no transition${transitionId ? ` '${transitionId}'` : ""} in manifest.contributes`);
    const fn = resolve(module, t.entry, "transition");
    const host = createHost(manifest, world);
    const active = world.layers.find((l) => l.id === world.activeLayerId);
    if (!active)
        throw new AzpError("world has no active layer");
    const ctx = { ...host, from: input.from, to: input.to, progress: input.progress, target: { id: active.id } };
    await fn(ctx);
}
