// @azphalt/runtime-reference/dist/world.js
/**
 * Allocate a zeroed `width × height` RGBA bitmap, optionally filled with a solid `fill` color.
 * `depth` defaults to 8 (a `Uint8ClampedArray`); pass 16 for a `Uint16Array`. `fill` channels are
 * written verbatim, so a 16-bit fill takes 16-bit channel values.
 */
export function blankBitmap(width, height, fill, depth = 8) {
    const data = depth === 16 ? new Uint16Array(width * height * 4) : new Uint8ClampedArray(width * height * 4);
    if (fill) {
        for (let i = 0; i < data.length; i += 4) {
            data[i] = fill.r;
            data[i + 1] = fill.g;
            data[i + 2] = fill.b;
            data[i + 3] = fill.a;
        }
    }
    return depth === 16 ? { data, width, height, depth } : { data, width, height };
}
/** Build a single-layer {@link World} — the usual starting point for exercising an extension. */
export function createWorld(opts = {}) {
    const width = opts.width ?? 4;
    const height = opts.height ?? 4;
    const layer = {
        id: "layer-0",
        name: "Layer 0",
        bitmap: blankBitmap(width, height),
        opacity: 1,
        blendMode: "normal",
    };
    return {
        canvas: { width, height, dpi: opts.dpi ?? 72 },
        layers: [layer],
        activeLayerId: layer.id,
        selection: null,
        color: { active: opts.activeColor ?? { r: 0, g: 0, b: 0, a: 255 }, palette: opts.palette ?? [] },
        params: opts.params ?? {},
        assets: opts.assets ?? {},
        time: { currentMs: opts.time?.currentMs ?? 0, durationMs: opts.time?.durationMs ?? 0, fps: opts.time?.fps ?? 30 },
        audio: opts.audio ?? null,
        redraws: 0,
    };
}
